src/raftstore/store/fsm/batch.rs

```diff
@@ -0,0 +1,575 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This is the core implementation of a batch system. Generally there will be two
+//! different kind of FSMs in TiKV's FSM system. One is normal FSM, which usually
+//! represents a peer, the other is control FSM, which usually represents something
+//! that controls how the former is created or metrics are collected.
+
+// TODO: remove this
+#![allow(dead_code)]
+
+use super::router::{BasicMailbox, Router};
+use crossbeam::channel::{self, SendError, TryRecvError};
+use std::borrow::Cow;
+use std::cmp;
+use std::thread::{self, JoinHandle};
+use util::mpsc;
+
+/// `FsmScheduler` schedules `Fsm` for later handles.
+pub trait FsmScheduler {
+    type Fsm: Fsm;
+
+    /// Schedule a Fsm for later handles.
+    fn schedule(&self, fsm: Box<Self::Fsm>);
+    /// Shutdown the scheduler, which indicates that resouces like
+    /// background thread pool should be released.
+    fn shutdown(&self);
+}
+
+/// A Fsm is a finite state machine. It should be able to be notified for
+/// updating internal state according to incomming messages.
+pub trait Fsm {
+    type Message: Send;
+
+    fn is_stopped(&self) -> bool;
+
+    /// Set a mailbox to Fsm, which should be used to send message to itself.
+    fn set_mailbox(&mut self, mailbox: Cow<BasicMailbox<Self>>)
+    where
+        Self: Sized;
+    /// Take the mailbox from Fsm. Implementation should ensure there will be
+    /// no reference to mailbox after calling this method.
+    fn take_mailbox(&mut self) -> Option<BasicMailbox<Self>>
+    where
+        Self: Sized;
+}
+
+/// A unify type for FSMs so that they can be sent to channel easily.
+enum FsmTypes<N, C> {
+    Normal(Box<N>),
+    Control(Box<C>),
+    Empty,
+}
+
+// A macro to introduce common definition of scheduler.
+macro_rules! impl_sched {
+    ($name:ident) => {
+        pub struct $name<N, C> {
+            sender: channel::Sender<FsmTypes<N, C>>,
+        }
+
+        impl<N, C> Clone for $name<N, C> {
+            #[inline]
+            fn clone(&self) -> $name<N, C> {
+                $name {
+                    sender: self.sender.clone(),
+                }
+            }
+        }
+    };
+}
+
+impl_sched!(NormalScheduler);
+impl_sched!(ControlScheduler);
+
+impl<N: Fsm, C> FsmScheduler for NormalScheduler<N, C> {
+    type Fsm = N;
+
+    #[inline]
+    fn schedule(&self, fsm: Box<N>) {
+        match self.sender.send(FsmTypes::Normal(fsm)) {
+            Ok(()) => return,
+            // TODO: use debug instead.
+            Err(SendError(FsmTypes::Normal(fsm))) => warn!("failed to schedule fsm {:p}", fsm),
+            _ => unreachable!(),
+        }
+    }
+
+    fn shutdown(&self) {
+        // TODO: close it explicitly once it's supported.
+        for _ in 0..100 {
```

Comment:
What does "100" mean?

Reply:
Magic number. Any number larger than the pool size is enough.

Reply:
Please add comments for the magic number.